File: treebylevels_gpt_opt.py
from collections import deque

class Node:
    def __init__(self, L, R, n):
        self.left = L
        self.right = R
        self.value = n
        


def tree_by_levels(root):
    if root is None:
        return []

    result, queue = [], deque([root])
    append, popleft = result.append, queue.popleft
    extend = queue.extend

    while queue:
        node = popleft()
        append(node.value)
        extend(child for child in (node.left, node.right) if child)

    return result


tree_by_levels(None)

tree_by_levels(Node(Node(None, Node(None, None, 4), 2), Node(Node(None, None, 5), Node(None, None, 6), 3), 1))

File: treebylevels_gpt_opt.js
class Node { 
  constructor(value, left = null, right = null) {
    this.value = value;
    this.left  = left;
    this.right = right;
  }
}

function treeByLevels(root) {
  const result = [];
  const queue = root ? [root] : [];

  for (let i = 0; i < queue.length; i++) {
    const { value, left, right } = queue[i];
    result.push(value);
    left  && queue.push(left);
    right && queue.push(right);
  }

  return result;
}

treeByLevels(null)

const treeOne =
  new Node(2,
    new Node(8,
      new Node(1),
      new Node(3)
    ),
    new Node(9,
      new Node(4),
      new Node(5)
    )
  );

treeByLevels(treeOne)

const treeTwo =
  new Node(1,
    new Node(8,
      null,
      new Node(3)
    ),
    new Node(4,
      null,
      new Node(5,
        null,
        new Node(7)
      )
    )
);

treeByLevels(treeTwo)